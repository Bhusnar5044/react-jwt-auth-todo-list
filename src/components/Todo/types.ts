import { ITodo } from '@context/types';

export interface Props {
    todo?: ITodo[];
    id?: number;
}
