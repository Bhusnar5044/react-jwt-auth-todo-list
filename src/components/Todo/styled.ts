import { Button } from '@components/Lib';
import { theme } from '@theme';
import { styled } from '@utils';

export const StyledContainer = styled.div`
    width: 100%;
    display: flex;
    justify-content: center;
    button {
        background: none;
        border: none;
        padding: 0;
        margin: 0;
        outline: none;
        box-shadow: none;
        cursor: pointer;
    }

    img,
    svg {
        display: block;
        margin: 0;
        max-width: 100%;
    }
`;

export const TodoContainer = styled.div`
    position: relative;
    width: 100%;
    width: 90%;
    border-radius: 0.5rem;
    background-color: #fff;
    box-shadow: rgba(100, 100, 111, 0.2) 0px 7px 29px 0px;
    text-align: center;
    overflow: hidden;
`;

export const TodoHeader = styled.div`
    padding: 2rem;
    background-color: #ccc;
    background-image: linear-gradient(135deg, #79f1a4 10%, #0e5cad 100%);
`;

export const StyledInput = styled.input`
    display: block;
    padding: 1rem;
    margin: 0;
    outline: none;
    color: ${theme.colors.info.main};
    height: 2rem;
    width: 90.8%;
    border: none;
    font-size: 1.4rem;
    font-weight: 700;
    padding-right: 0.75rem;
`;

export const StyledButton = styled(Button)`
    background-color: transparent;
    color: ${theme.colors.info.dark};
    font-size: 2rem;
    &:hover {
        color: ${theme.colors.info.main};
        background-color: transparent;
        box-shadow: unset;
    }
`;

export const TodoForm = styled.div`
    position: relative;
    display: flex;
    width: 100%;
    align-items: center;
    text-align: left;
    border-bottom: 1px solid #e9e9e9;
    padding-left: 1.5rem;
    padding-right: 1rem;
`;

export const TodoBody = styled.div`
    position: relative;
    padding: 2rem 2.5rem;
    height: 420px;
    overflow: auto;
`;
