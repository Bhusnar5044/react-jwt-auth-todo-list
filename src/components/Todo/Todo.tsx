import { Text } from '@components/Lib';
import { useTodoContext } from '@context/TodoProvider';
import { FC, memo, useEffect, useRef, useState } from 'react';
import List from './List';
import { StyledButton, StyledContainer, StyledInput, TodoBody, TodoContainer, TodoForm, TodoHeader } from './styled';

export const Todo: FC = memo(() => {
    const { todo: tasks, setTodo: setTasks } = useTodoContext();
    const [remainingTasks, setRemainingTasks] = useState(0);
    const addTaskInput = useRef<HTMLInputElement>(null);

    useEffect(() => {
        setRemainingTasks(tasks?.filter((taskItem) => taskItem.done === false).length || 0);
    }, [tasks]);

    function addTask() {
        if (addTaskInput?.current) {
            const newTask = addTaskInput?.current.value;
            if (newTask && !tasks?.some((task) => task.text === newTask)) {
                const newTaskObj = {
                    text: newTask,
                    done: false,
                };
                setTasks?.((tasks) => {
                    return [...tasks, newTaskObj];
                });
                addTaskInput.current.value = '';
                addTaskInput.current.focus();
            }
        }
    }

    function handleKey({ keyCode, target }: { keyCode: number; target: any }) {
        if (keyCode === 13) {
            addTask();
        } else if (addTaskInput?.current) {
            if (!tasks?.some((task) => task.text === target.value)) {
                addTaskInput?.current.classList.remove('invalid');
            } else {
                addTaskInput?.current.classList.add('invalid');
            }
        }
    }

    return (
        <StyledContainer className="main">
            <TodoContainer className="todo">
                <TodoHeader className="todo-header">
                    <Text textVariant="h3">Todo List</Text>
                    <Text textVariant="h5">
                        {tasks?.length && remainingTasks === 0 ? (
                            'All done! =D'
                        ) : (
                            <>
                                You have <strong>{remainingTasks}</strong> of <strong>{tasks?.length}</strong> tasks
                                remaining
                            </>
                        )}
                    </Text>
                </TodoHeader>
                <TodoForm className="todo-form">
                    <StyledInput ref={addTaskInput} type="text" placeholder="Add task..." onKeyUp={handleKey} />
                    <StyledButton onClick={addTask}>
                        <Text textVariant="h4">+</Text>
                    </StyledButton>
                </TodoForm>
                <TodoBody className="todo-body">
                    <List />
                </TodoBody>
            </TodoContainer>
        </StyledContainer>
    );
});

Todo.displayName = 'Todo';
