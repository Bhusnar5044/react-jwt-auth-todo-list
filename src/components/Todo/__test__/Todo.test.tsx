import { render } from '@test-utils';
import { Todo } from '../Todo';

describe('Todo', () => {
    it('should render properly', () => {
        const { container } = render(<Todo />);
        expect(container).toMatchSnapshot();
    });
});
