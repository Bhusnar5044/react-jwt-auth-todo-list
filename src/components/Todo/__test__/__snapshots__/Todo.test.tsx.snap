// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo should render properly 1`] = `
.c7 {
  position: relative;
  display: inline-block;
  outline: none;
  cursor: pointer;
  text-align: center;
  border: unset;
  padding: 1rem 2rem;
  color: #fff;
  background: #008085;
  border-radius: 0.8rem;
  font-weight: 600;
  font-size: 1.4rem;
  line-height: 2.2rem;
  box-shadow: rgba(0, 0, 0, 0.05) 0px 1px 0px 0px,rgba(0, 0, 0, 0.2) 0px -1px 0px 0px inset;
  opacity: 1;
}

.c7:hover {
  background: #4caf50;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2),0 6px 20px 0 rgba(0, 0, 0, 0.19);
}

.c7.loading {
  cursor: not-allowed;
  pointer-events: none;
  opacity: 0.7;
}

.c7.loading::after {
  content: '';
  position: absolute;
  width: 2rem;
  height: 2rem;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: auto;
  border: 0.4rem solid transparent;
  border-top-color: #1565c0;
  border-right-color: #1565c0;
  border-radius: 50%;
  animation: gNWkkb 1s ease infinite;
}

.c3 {
  margin: 0;
  color: inherit;
  font-size: 2.4rem;
  font-weight: 400;
  letter-spacing: -0.04rem;
  line-height: 3.2rem;
  text-align: initial;
}

.c4 {
  margin: 0;
  color: inherit;
  font-size: 1.2rem;
  font-weight: 700;
  letter-spacing: 0.12rem;
  line-height: 2.0rem;
  text-align: initial;
}

.c0 {
  width: 100%;
  display: flex;
  justify-content: center;
}

.c1 {
  position: relative;
  width: 100%;
  width: 90%;
  border-radius: 0.5rem;
  background-color: #fff;
  box-shadow: rgba(100, 100, 111, 0.2) 0px 7px 29px 0px;
  text-align: center;
  overflow: hidden;
}

.c2 {
  padding: 2rem;
  background-color: #ccc;
  background-image: linear-gradient(135deg, #79f1a4 10%, #0e5cad 100%);
}

.c6 {
  display: block;
  padding: 1rem;
  margin: 0;
  outline: none;
  color: #495057;
  height: 2rem;
  width: 90.8%;
  border: none;
  font-size: 1.4rem;
  font-weight: 700;
  padding-right: 0.75rem;
}

.c8 {
  background: none;
  border: none;
  padding: 0;
  margin: 0;
  outline: none;
  box-shadow: none;
  background-color: transparent;
  color: #212529;
  font-size: 2rem;
}

.c8:hover {
  color: #495057;
  background-color: transparent;
  box-shadow: unset;
}

.c5 {
  position: relative;
  display: flex;
  width: 100%;
  align-items: center;
  text-align: left;
  border-bottom: 1px solid #e9e9e9;
  padding-left: 1.5rem;
  padding-right: 1rem;
}

.c9 {
  position: relative;
  padding: 2rem 1.5rem;
  height: fit-content;
  overflow: auto;
}

.c10 {
  margin-top: 2rem;
}

.c10 p {
  margin: 0;
  margin-bottom: 1rem;
}

.c10 svg {
  display: inline-block;
  width: 50px;
  opacity: 0.5;
}

<div>
  <div
    class="c0 main"
  >
    <div
      class="c1 todo"
    >
      <div
        class="c2 todo-header"
      >
        <h3
          class="c3"
          textalign="initial"
          textcolor="inherit"
          textvariant="h3"
        >
          Todo List
        </h3>
        <h5
          class="c4"
          textalign="initial"
          textcolor="inherit"
          textvariant="h5"
        >
          You have 
          <strong>
            0
          </strong>
           of
           
          <strong />
           tasks remaining
        </h5>
      </div>
      <div
        class="c5 todo-form"
      >
        <input
          class="c6"
          placeholder="Add task..."
          type="text"
        />
        <button
          class="c7 c8"
          type="button"
        >
          +
        </button>
      </div>
      <div
        class="c9 todo-body"
      >
        <div
          class="c10 empty"
        >
          <p>
            <svg
              viewBox="0 0 21 21"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fill-rule="evenodd"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                transform="translate(2.5 4.5)"
              >
                <path
                  d="m3.65939616 0h8.68120764c.4000282 0 .7615663.23839685.9191451.6060807l2.7402511 6.3939193v4c0 1.1045695-.8954305 2-2 2h-12c-1.1045695 0-2-.8954305-2-2v-4l2.74025113-6.3939193c.15757879-.36768385.51911692-.6060807.91914503-.6060807z"
                />
                <path
                  d="m0 7h4c.55228475 0 1 .44771525 1 1v1c0 .55228475.44771525 1 1 1h4c.5522847 0 1-.44771525 1-1v-1c0-.55228475.4477153-1 1-1h4"
                />
              </g>
            </svg>
          </p>
          <p>
            <b>
              Empty list
            </b>
          </p>
          <p>
            Add a new task above
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
